docs/api.rst

```diff
@@ -0,0 +1,287 @@
+===========
+Treelib API
+===========
+
+Library for manipulating trees in Python made up of dicts and lists.
+
+
+Goals
+=====
+
+The primary goal of this library is to make it less unwieldy to manipulate trees
+made up of Python dicts and lists.
+
+For example, say we want to get a value deep in the tree. we could do this::
+
+  value = tree['a']['b']['c']
+
+
+That'll throw a ``KeyError`` if any of those bits are missing. So you could
+handle that::
+
+  try:
+      value = tree['a']['b']['c']
+  except KeyError:
+      value = None
+
+
+Alternatively, you could do this::
+
+  value = tree.get('a', {}).get('b', {}).get('c': None)
+
+
+These work, but both are unwieldy especially if you're doing this a lot.
+
+Similarly, setting things deep is also unenthusing::
+
+  tree['a']['b']['c'] = 5
+
+
+The safer form is this::
+
+  tree.setdefault('a', {}).setdefault('b', {})['c'] = 5
+
+
+This library aims to make sane use cases for tree manipulation easier to read
+and think about.
+
+
+Paths
+=====
+
+A path is a string specifying a period-delimited list of edges. Edges can be:
+
+1. a key (for a dict)
+2. an index (for a list)
+
+Example paths::
+
+  a
+  a.[1].foo_bar.Bar
+  a.b.[-1].Bar
+
+
+Paths can be composed using string operations since they're just strings.
+
+FIXME(willkg): Add diagram showing a tree with edges specified by a path.
+
```

Comment:
Paths are based on, but not as expressive as, jq filters: https://stedolan.github.io/jq/tutorial/